source/lac/petsc_parallel_vector.cc

```diff
@@ -236,11 +236,24 @@ namespace PETScWrappers
 
     void
     Vector::reinit(
-      const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner)
+      const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner,
+      const bool                                                make_ghosted)
     {
-      this->reinit(partitioner->locally_owned_range(),
-                   partitioner->ghost_indices(),
-                   partitioner->get_mpi_communicator());
+      if (make_ghosted)
+        {
+          Assert(partitioner->ghost_indices_initialized(),
+                 ExcMessage("You asked to create a ghosted vector, but the "
+                            "partitioner does not provide ghost indices."));
+
+          this->reinit(partitioner->locally_owned_range(),
+                       partitioner->ghost_indices(),
+                       partitioner->get_mpi_communicator());
+        }
+      else
+        {
+          this->reinit(partitioner->locally_owned_range(),
+                       partitioner->get_mpi_communicator());
+        }
```

Comment:
Why not:
```suggestion
      if (partitioner->ghost_indices_initialized())
        {
          this->reinit(partitioner->locally_owned_range(),
                       partitioner->ghost_indices(),
                       partitioner->get_mpi_communicator());
        }
      else
        {
          this->reinit(partitioner->locally_owned_range(),
                       partitioner->get_mpi_communicator());
        }
```

Reply:
With the `bool` parameter you can still create a non-ghosted vector out of a partitioner with ghost elements.

With your change, I would need two different partitioners: one with and one without ghost elements. That would work with me as well, but the question is whether we want that. I haven't worked much with the partitioner class: Is it common in user-code to have multiple partitioners like this?

Reply:
After thinking and fiddling about this for a little longer, I find that your suggestion is the way to go.The reinit function should correspond to whatever the partitioner provides. I will adjust this PR and the documentation.

Reply:
...well I thought so. With the new suggestion I encountered some strange behavior.

I initialize the `system_rhs` and pass it to `AffineConstraints::distribute_local_to_global(cell_matrix, cell_rhs, local_dof_indices, system_matrix, system_rhs)` which causes problems:
- Initialized with ghost elements, it fails when `system_rhs` is a PETSc/Trilinos Vector, and succeeds when it is a LA::distributed::Vector
- Initialized without ghost elements, it succeeds when `system_rhs` is a PETSc/Trilinos Vector, and fails when it is a LA::distributed::Vector

I'm puzzled. I feel tempted to remove the last commit in this PR. What do you think?

Reply:
I think we indeed need to enable the user to control where to take into account the ghosts, so the current solution with the additional flag seems right to me.